calibre-web: init at 0.6.11

pborzenkov opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

The PR adds `calibre-web` package, which is a Web frontend for Calibre ebook management software.

There are two patches applied to `calibre-web` in order to make it more
NixOS friendly:

  - default-logger.patch switches default log output to /dev/stdout, as
    otherwise calibre-web tries to open a file relative to its location,
    which it can't do as the store is read-only. It's not possible to
    pass log file location via command line flags.

  - run-migrations.patch adds a cli flag that instructs calibre-web to
    run DB migrations and exit. As almost all config parameters are
    configured via UI in sqlite3 DB, this patch allows the DB to be
    pre-created so it can be updated by systemd pre-start script later.
    Thus, allowing calibre-web to be configured declaratively.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

pborzenkov committed:
calibre-web: init at 0.6.11

There are two patches applied to 'calibre-web' in order to make it more
NixOS friendly:

  - default-logger.patch switches default log output to /dev/stdout, as
    otherwise calibre-web tries to open a file relative to its location,
    which it can't do as the store is read-only. It's not possible to
    pass log file location via command line flags.

  - run-migrations.patch adds an env var __RUN_MIGRATIONS_AND_EXIT that,
    when set, instructs calibre-web to run DB migrations and exit. As
    almost all config parameters are configured via UI in sqlite3 DB,
    this patch allows the DB to be pre-created so it can be updated by
    systemd pre-start script later.  Thus, allowing calibre-web to be
    configured declaratively.

r-rmcgibbo commented:
Result of `nixpkgs-review pr 113994` at dfcb7de9 run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li><a href="https://gist.github.com/4a6834f5b6cf79b88cfcbbe18113b68b">calibre-web</a></li></li>
  </ul>
</details>
<details>
  <summary>2 suggestions:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-patch-comment.md">missing-patch-comment</a>

Please add a comment on the line above, explaining the purpose of this patch.
Near pkgs/servers/calibre-web/default.nix:24:5:
```
   |
24 |     ./default-logger.patch
   |     ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-patch-comment.md">missing-patch-comment</a>

Please add a comment on the line above, explaining the purpose of this patch.
Near pkgs/servers/calibre-web/default.nix:25:5:
```
   |
25 |     ./db-migrations.patch
   |     ^
```
</li>
  </ul>
</details>

humancalico commented on maintainers/maintainer-list.nix:
```diff
@@ -7203,6 +7203,12 @@
     githubId = 157610;
     name = "Piotr Bogdan";
   };
+  pborzenkov = {
+    email = "pavel@borzenkov.net";
+    github = "pborzenkov";
+    githubId = 434254;
+    name = "Pavel Borzenkov";
+  };
```
Please add yourself as a maintainer in a separate commit with the message `maintainers: add pborzenkov`.

pborzenkov replied:
Done!

SuperSandro2000 commented on pkgs/servers/calibre-web/default.nix:
```diff
@@ -0,0 +1,67 @@
+
+python3.pkgs.buildPythonApplication rec {
+  pname = "calibre-web";
+  version = "0.6.11";
+  src = fetchFromGitHub {
```
```suggestion
  version = "0.6.11";

  src = fetchFromGitHub {
```

SuperSandro2000 commented on pkgs/servers/calibre-web/default.nix:
```diff
@@ -0,0 +1,67 @@
+
+  prePatch = ''
+    substituteInPlace setup.cfg \
+        --replace "requests>=2.11.1,<2.25.0" "requests>=2.11.1,<=2.25.1" \
+        --replace "cps = calibreweb:main" "calibre-web = calibreweb:main"
```
```suggestion
      --replace "requests>=2.11.1,<2.25.0" "requests>=2.11.1,<=2.26.0" \
      --replace "cps = calibreweb:main" "calibre-web = calibreweb:main"
```

SuperSandro2000 commented on pkgs/servers/calibre-web/default.nix:
```diff
@@ -0,0 +1,67 @@
+    ./default-logger.patch
+    # DB migrations adds an '-m' CLI option that instructs calibre-web to run DB migrations and exit. This is gonna be
+    # used to configure calibre-web declaratively, as most of its configuration parameters are stored in the DB.
+    ./db-migrations.patch
```
I don't think packaging should add new options. Is there an upstream commit with those changes we could fetch?

pborzenkov replied:
> I don't think packaging should add new options.

I agree. I was rather reluctant to add it. But the only way to configure `calibre-web` is through the sqlite DB. So if we want to have a declarative module, and I'd like to have one, we need to have a way to ensure that DB exists even before the first start.

The only option that can be set is the listening port (via env var). This variable is only consulted on the first start when there is no DB. So even that can't be put into module.

There is an upstream issue (https://github.com/janeczku/calibre-web/issues/1856) with a proposal to add configuration file support. But before it gets implemented (if it gets implemented at all), there is basically no way to create a module.

If you really don't like the patch, I can drop it and settle with configuring `calibre-web` via UI.

> Is there an upstream commit with those changes we could fetch?

No.

SuperSandro2000 commented on pkgs/servers/calibre-web/default.nix:
```diff
@@ -0,0 +1,67 @@
+    Wand
+  ];
+
+  doCheck = false;
```
Please add a comment why tests are disabled. Can we add an pythonImportsCheck instead?

pborzenkov replied:
Actually, it's my mistake. I disabled them while writing the derivation and forgot to enable them back. Will do.

SuperSandro2000 commented on pkgs/servers/calibre-web/default.nix:
```diff
@@ -0,0 +1,67 @@
+
+  doCheck = false;
+
+  meta = with lib;
+    {
+      description = "Web app for browsing, reading and downloading eBooks stored in a Calibre database";
+      maintainers = with maintainers; [ pborzenkov ];
+      homepage = "https://github.com/janeczku/calibre-web";
+      license = licenses.gpl3;
+      platforms = platforms.all;
+    };
```
```suggestion
  meta = with lib; {
    description = "Web app for browsing, reading and downloading eBooks stored in a Calibre database";
    maintainers = with maintainers; [ pborzenkov ];
    homepage = "https://github.com/janeczku/calibre-web";
    license = licenses.gpl3;
    platforms = platforms.all;
  };
```

pborzenkov committed:
nixos/calibre-web: init module

pborzenkov commented:
@SuperSandro2000  I've addressed your review comments. I've also dropped a command line option and use and env var instead, so there won't be a collision in case upstream introduces `-m` cli option.

I've also added a NixOS module to the PR.

Ambroisie commented on nixos/modules/services/web-apps/calibre-web.nix:
```diff
@@ -0,0 +1,162 @@
+        };
+      };
+
+      dataDir = mkOption {
+        type = types.str;
+        default = "calibre-web";
+        description = ''
+          The directory below <filename>/var/lib</filename> where Calibre-Web stores its data.
+        '';
+      };
```
Shouldn't the full path be configurable? Not that I would ever use anything outside of `/var/lib` anyway, but other modules allow for an arbitrary path in this case (e.g: `sonarr`).

symphorien replied:
One can argue that it's easy to make a bind mount if necessary.

pborzenkov replied:
> Shouldn't the full path be configurable?

I've looked through various modules and they use both approaches. Some use systemd `StateDirectory`, which implies that the directory is below `/var/lib`; some allow to configure full path. The benefit of `StateDirectory` is that systemd ensures that the directory exists and has correct user/group/permissions before starting the service. So I like it better.

Ambroisie commented:
Is there anything else left to do on this? It seems that the requested changes have been done even though github does not display them as so.

SuperSandro2000 commented on pkgs/servers/calibre-web/default.nix:
```diff
@@ -0,0 +1,66 @@
+    description = "Web app for browsing, reading and downloading eBooks stored in a Calibre database";
+    maintainers = with maintainers; [ pborzenkov ];
+    homepage = "https://github.com/janeczku/calibre-web";
+    license = licenses.gpl3;
```
gpl3Only or gpl3Plus?

pborzenkov replied:
The README doesn't have any info, but some of the files in the repo has the following phrase in the header comment:

> either version 3 of the License, or (at your option) any later version.

So I'll update the PR with `gpl3Plus`.

SuperSandro2000 commented:
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 113994` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
1 package blacklisted:
</summary>
<ul>
<li>tests.nixos-functions.nixos-test</li>
</ul>
</details>
<details>
<summary>
1 package built:
</summary>
<ul>
<li>calibre-web</li>
</ul>
</details>

The following issues got detected with the above build packages.
Please fix at least the ones listed with your changed packages:

<details>
calibre-web:
<details>


warning: no-python-tests
Test runner could not discover any test cases: ‘Ran 0 tests in 0.000s’
See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/no-python-tests.md
</details>
</details>

pborzenkov commented:
> calibre-web:
> warning: no-python-tests
> Test runner could not discover any test cases: ‘Ran 0 tests in 0.000s’
> See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/no-python-tests.md

Will disable the tests are upstream repo doesn't have any.

symphorien commented on pkgs/servers/calibre-web/default.nix:
```diff
@@ -0,0 +1,69 @@
+    Wand
+  ];
+
+  meta = with lib; {
```
you can add passthru.tests

symphorien commented on nixos/modules/services/web-apps/calibre-web.nix:
```diff
@@ -0,0 +1,162 @@
+        '';
+      };
+
+      settings = {
```
if the option is called settings, it should respect RFC42 https://github.com/NixOS/rfcs/blob/45b76f20add8d7e0e1d0bc0eed9357b067c899a6/rfcs/0042-config-option.md

It looks like the format for the configuration file is INI, so it should be easy.

pborzenkov replied:
I'll have a look at the RFC. But just FYI calibre-web doesn't have a configuration file. The settings listed in the module are written into sqlite DB by systemd pre-exec script (it's the only way to configure calibre-web).

symphorien replied:
Oh indeed I had not seen the string constructed line 111 was fed to sqlite3. Well in this case I'd suggest renaming the option to something else than settings, just to avoid suggesting this is a RFC42-style settings option.

symphorien reviewed (changes requested):
I had a quick look, I plan to try and execute the module later.

Ambroisie commented on nixos/modules/misc/ids.nix:
```diff
@@ -347,6 +347,7 @@ in
       #mailman = 316;  # removed 2019-08-30
       zigbee2mqtt = 317;
       # shadow = 318; # unused
+      calibre-web = 319;
```
Aren't hard-coded uids for services supposed to stop being used?

SuperSandro2000 replied:
Yes, you are right. Good catch!

pborzenkov replied:
> Aren't hard-coded uids for services supposed to stop being used?

@SuperSandro2000 Ok, will remove. What is the correct way to specify user/group in this case? `calibre-web` needs r/w access to Calibre DB. What should I do in this case? Should I use systemd's `DynamicUser` and add provide some `SupplimentaryGroups` that have r/w access to the DB?

Ambroisie replied:
Can't you just not specify an ID for the `calibre-web` user and group. NixOS should pick an ID automatically.

You might need to specify that it is a system user.

Ambroisie replied:
See this paragraph from the linked RFC.

> Otherwise let NixOS assign dynamic persistent ids, which happens automatically when users.users.<name?>.uid/users.groups.<name?>.gid is not set. For users, users.users.<name?>.isSystemUser should be set so that only uids under 1000 are used by NixOS services.

pborzenkov replied:
> Can't you just not specify an ID for the `calibre-web` user and group. NixOS should pick an ID automatically.

Sure I can. But this automatically generated user won't have write permissions to the original Calibre DB.

Ambroisie commented on nixos/modules/misc/ids.nix:
```diff
@@ -649,6 +650,7 @@ in
       #mailman = 316;  # removed 2019-08-30
       zigbee2mqtt = 317;
       shadow = 318;
+      calibre-web = 319;
```
Ditto for gids.

Ambroisie commented:
I'm referencing [this RFC](https://github.com/NixOS/rfcs/blob/master/rfcs/0052-dynamic-ids.md) for the IDs. Is there something that I am missing with this approach (for example not being able to configure the group/user under which the service must be run)?

symphorien reviewed (changes requested):
I left a few secondary inline comments.

I tried to run this PR in a nixos-rebuild build-vm vm. I set
```
    services.calibre-web.enable = true;
    services.calibre-web.options.calibreLibrary = "/home/alice/tmp/calibre";
    services.calibre-web.user = "alice";
    services.calibre-web.group = "users";
```
and manually copied the output of `calibredb export --all` to the vm at `/home/alice/tmp/calibre`.
When I navigate to 127.0.0.1:8083 calibreweb asks for the location of the db.
So it seems that the setting for the calibre library is ignored.

pborzenkov commented:
> When I navigate to 127.0.0.1:8083 calibreweb asks for the location of the db.
> So it seems that the setting for the calibre library is ignored.

Did you restart `calibre-web` service after you copied your DB to the VM? This is an unfortunate peculiarity in `calibre-web` - it simply resets the setting if it can't find valid Calibre DB at the specified path. This is exactly the reason why I restart it in the tests.

symphorien commented:
> Did you restart calibre-web service after you copied your DB to the VM? 

I rebooted the vm and same happens

>  it simply resets the setting if it can't find valid Calibre DB at the specified path

hum that's nasty -- could we check (for example with calibredb, or simply with `test -d`) in preStart that the declaratively configured database is valid and if it is not make the service fail?
something like 
` test -d ${calibreLibrary} || { echo "invalid library at ${calibreLibrary}"; exit 1; }`

symphorien commented on nixos/modules/services/web-apps/calibre-web.nix:
```diff
@@ -0,0 +1,159 @@
+      listen = {
+        ip = mkOption {
+          type = types.str;
+          default = "127.0.0.1";
```
hum for some reason it does not listen on both ipv6 and ipv4 at the same time, and it does not accept several ip addresses with several -i options...

pborzenkov commented:
@symphorien 

> and manually copied the output of `calibredb export --all` to the vm at `/home/alice/tmp/calibre`.
> When I navigate to 127.0.0.1:8083 calibreweb asks for the location of the db.
> So it seems that the setting for the calibre library is ignored.

Ok, I've found what was the problem. `calibredb export` does not produce a Calibre database (e.g. there is not `metadata.db` file), it just applies metadata to all the books in the library and stores them at the specified location.

Calibre-Web requires `metadata.db` to be present at the specified path.

pborzenkov commented:
> hum that's nasty -- could we check (for example with calibredb, or simply with `test -d`) in preStart that the declaratively configured database is valid and if it is not make the service fail?
> something like
> ` test -d ${calibreLibrary} || { echo "invalid library at ${calibreLibrary}"; exit 1; }`

Done!

symphorien commented:
Nice finding !
I was able to go further. Now I get to the login page, but I have no credentials. I see there is an option to authenticate with a reverse proxy, but if it is an option then how do I configure users when it is false?

pborzenkov commented:
> Nice finding !
> I was able to go further. Now I get to the login page, but I have no credentials. I see there is an option to authenticate with a reverse proxy, but if it is an option then how do I configure users when it is false?

As per https://github.com/janeczku/calibre-web#quick-start the default credentials are:

> Default admin login:
> Username: admin
> Password: admin123

Even with reverse auth proxy users still need to be created in the calibre-web itself, it just allows to bypass login screen.
